packages/zone.js/lib/zone-impl.ts

```diff
@@ -1504,57 +1497,71 @@ export function initZone(): ZoneType {
   const symbolSetTimeout = __symbol__('setTimeout');
   const symbolPromise = __symbol__('Promise');
   const symbolThen = __symbol__('then');
+  // To prevent any breaking changes resulting from this change, given that
+  // it was already causing a significant number of failures in g3, we have hidden
+  // that behavior behind a global configuration flag. Consumers can enable this
+  // flag explicitly if they want the microtask queue to be drained as defined
+  // in the specification.
+  const enableNativeMicrotaskDraining = __symbol__('enable_native_microtask_draining');
   let _microTaskQueue: Task[] = [];
-  let _isDrainingMicrotaskQueue: boolean = false;
+  let _isDrainingMicrotaskQueue = false;
   let nativeMicroTaskQueuePromise: any;
 
   function nativeScheduleMicroTask(func: Function) {
-    if (!nativeMicroTaskQueuePromise) {
-      if (global[symbolPromise]) {
-        nativeMicroTaskQueuePromise = global[symbolPromise].resolve(0);
-      }
+    if (!nativeMicroTaskQueuePromise && global[symbolPromise]) {
+      nativeMicroTaskQueuePromise = global[symbolPromise].resolve(0);
     }
+
     if (nativeMicroTaskQueuePromise) {
-      let nativeThen = nativeMicroTaskQueuePromise[symbolThen];
-      if (!nativeThen) {
-        // native Promise is not patchable, we need to use `then` directly
-        // issue 1078
-        nativeThen = nativeMicroTaskQueuePromise['then'];
-      }
-      nativeThen.call(nativeMicroTaskQueuePromise, func);
+      const thenFn = nativeMicroTaskQueuePromise[symbolThen] ?? nativeMicroTaskQueuePromise['then']; // fallback for non-patchable Promise
+      // Use the resolved native promise to schedule the microtask
+      thenFn.call(nativeMicroTaskQueuePromise, func);
     } else {
+      // Fallback to setTimeout if native promise is unavailable
       global[symbolSetTimeout](func, 0);
     }
   }
 
   function scheduleMicroTask(task?: MicroTask) {
-    // if we are not running in any task, and there has not been anything scheduled
-    // we must bootstrap the initial task creation by manually scheduling the drain
-    if (_numberOfNestedTaskFrames === 0 && _microTaskQueue.length === 0) {
-      // We are not running in Task, so we need to kickstart the microtask queue.
-      nativeScheduleMicroTask(drainMicroTaskQueue);
+    const isNativeDrainingEnabled = global[enableNativeMicrotaskDraining];
+    const shouldDrainWithNative = isNativeDrainingEnabled && !_isDrainingMicrotaskQueue;
```

Comment:
```suggestion
    const shouldDrain = _microTaskQueue.length === 0 && (isNativeDrainingEnabled ? !_isDrainingMicrotaskQueue : _numberOfNestedTaskFrames === 0);
```

and really it seems the old behavior should also not schedule a drain if it’s already draining, right? Maybe we can try that change in a separate PR.